macros/graph/plotly3D.pl

```diff
@@ -0,0 +1,626 @@
+ranging from 0 to 1. For example the default RdBu is the following colorset
+(note this must be a string since the array is passed to javascript):
+
+    "[[0, 'rgb(5,10,172)'], [0.35, 'rgb(106,137,247)'],
+      [0.5, 'rgb(190,190,190)'], [0.6, 'rgb(220,170,132)'],
+      [0.7, 'rgb(230,145,90)'], [1, 'rgb(178,10,28)']]"
+
+A colorset can have any number of color points between 0 and 1. To make
+the plot a single color, set the color for 0 and 1 to be the same:
+
+    "[[0, 'rgb(0,200,0)'], [1, 'rgb(0,200,0)']]"
+
+=head1 FUNCTION TYPES
+
+The functions to generate the plot can be either mathematical, JavaScript,
+Perl, or raw data, and this can be controlled using the funcType => type
+option in addCurve or addSurface methods. The valid types are:
+
+  jsmd    This is the default type, in which the functions are converted
+          from math formulas into JavaScript functions to generate the
+          plot. This should accept standard mathematical notation with
+          some exceptions: Multiplication must be an explicit "*".
+          "ucos(v)" is not be accepted, but "u*cos(v)" will. JavaScript
+          considers "-u^2" not well defined, instead use "-(u^2)".
+
+  js      The functions are interpreted as raw JavaScript functions. The
+          functions will be passed the defined variables and return a
+          single value.
+
+  perl    The functions are interpreted as Perl subroutines. The functions
+          will be passed the appropriate number of inputs, and return a
+          single value.
+
+  data    The functions are interpreted as a nested array of data points to
+          be sent directly to plotly to plot. This is useful for static plots
+          in which the points do not need to be generated each time.
+
+=head1 Graph3D OPTIONS
+
+Create a graph object: $graph = Graph3D(option => value)
+The valid options are:
+
+  height      The height and width of the div containing the graph.
+  width
+
+  title       Graph title to print above the graph.
+
+  style       CSS style to style the div containing the graph.
+
+  bgcolor     The background color of the graph.
+
+  image       Image filename to be used in hardcopy TeX output.
+              If no image is provided, the hardcopy TeX output
+              has a message that image must be viewed online.
+
+  tex_size    Size of image in hardcopy TeX output as scale factor from 0 to 1000.
+              1000 is 100%, 500 is 50%, etc.
+
+  tex_border  Put (1) or don't put (0) a border around image in TeX output.
+
+  scene       Add a JavaScript scene configuration dictionary to the plotly layout.
+              Example: scene => 'aspectmode: "manual", aspectratio: {x: 1, y: 1, z: 1},
+              xaxis: { range: [0,2] }, yaxis: { range: [0,3] }, zaxis: { range: [1,4] }'
+              https://plotly.com/javascript/3d-axes/ for more examples.
+
+=cut
+
+sub _plotly3D_init {
+	ADD_JS_FILE('https://cdn.plot.ly/plotly-latest.min.js', 1);
```

Comment:
@somiaj:  This should not load `plotly` from a cdn.  It should be installed via npm with `npm install plotly.js-dist-min`.  The third-party-assets.json file should list the appropriate cdn for it.  It seems that would be `https://cdn.jsdelivr.net/npm/plotly.js-dist-min/plotly.min.js`.  That gives the system administrator the choice of serving it directly (which many prefer -- I certainly do) or serving it from a cdn.  It is also a bad idea to use `plotly-latest.min.js`.  Instead use a specific version.   You can use the actual latest version at this point for now, but this may cause problems later.  Always specify versions.

Reply:
I set the version to be 2.18.2, the latest version in `package.json` and updated the cdn to match that version. Any reason to use jsdelivr over the suggested CDN on the plotly.js github page of `https://cdn.plot.ly/plotly-2.18.2.js`. Assuming this is a more generic CDN vs one just for plot.ly. Also do I need to do anything with the `package-lock.json` file?

Reply:
For consistency it is best to use the jsdelivr one.  It will be the same as the one installed by npm.

Reply:
When you run `npm install plotly.js-dist-min` it will automatically add what is needed to the package-lock.json file.  You will need to commit that and include that in your pull request.

Reply:
By the way, hopefully you did not manually edit the `package.json` file to add `ploty`.  You should just run `npm install plotly.js-dist-min` as I said.  That will add what is needed to both the `package.json` file and the `package-lock.json` file.  By default it will use the latest version, which is 2.18.2.